### ISSUE TITLE
creating midi file

### ISSUE THREAD

**Issue by jisike** [moinejf/abc2svg#59]
abc2svg is the only lib that can render and playback the my multi-part abc file. is there anyway i can create a midi/hack something similar using your lib?

**Comment by moinejf** [comment]
I split the audio generation into small tools. For MIDI generation, just replace util/toaudio5.js by MIDI file creation.

**Comment by jisike** [comment]
could you please give me more detail

**Comment by moinejf** [comment]
I think the interface in util/toaudio5.js is clear enough.
You get this source, remove the code of the functions, and convert the play events to a binary MIDI stream.
Then, you concatenate util/play.js, util/toaudio.js and your code, and you load the resulting file at the place of play-1.js.

**Comment by bwl21** [comment]
I am also pulling the play events from the player in the callback get_abcmodel.

          abcplay.add(tsfirst, voice_tb)
          player_model = abcplay.clear()

where abcplay is an instance of AbcPlay.

This approach has problems for me:

1. in order to get the player events (player_model), I need an instance of `AbcPlay`, even if i do not really want to play. 
2. to get an instance of AbcPlay, i need an Audio context which is not available in node.
3. I cannot create an AbcPlay on the fly since every instance reserves an Audio context which I could not release. So I cannot do it more than six times. 
4. I have two components ABC2SVG and HarpnotePlayer (which plays the events). Now I had to pass the instance of AbcPlay from HarpnotePlayer to ABC2SVG which was pretty complicated.

Long story short:

Would it be possible to factor out the creation of the player events from AbcPlay to either another module (e.g. `AbcPlayerEvents` like AbcJSON) or to `AbcMIDI` respectively `Abc` (where parsing and get_abcmodel is done)

This would simplify the subsequent creation of a midi-file.

**Comment by moinejf** [comment]
As told previously, the play function has been split into toaudio (play event generation) and toaudio5 (html5 audio playing engine). The file play.js is just a wrapper around these two modules.
The API is simple enough:
- toaudio gets the ABC music representation and generates the play event list
- to audio5 gets the play event list and generates the music through the html5 audio interface
The play event list is described in toaudio5.
This permits:
- to replace toaudio5 by some other modules and generate music by some other mean (MIDI stream/file, csound file, ...)
- to call toaudio5 and play music  from some other format (MIDI file/device, lilypond/musicxml/pmw/cmn/guido/braillemusic/... music notations...)

**Comment by bwl21** [comment]
I take this as a confirmation the the approach of using ToAudio to get the player events is ok.
You mentioned that you want to change the player events from an array to a float32 array.

**Comment by moinejf** [comment]
All values in the play event array are numbers. Generating a standard array or a float32array does not change its usage.
